### PR TITLE
add filtering elements when get sub element of directory

### PR DESCRIPTION
**Pull request by YenguiSeddik**

### CODE REVIEW

File: src/main/java/org/gridsuite/explore/server/services/DirectoryService.java [hunk 1]
```diff
@@ -91,9 +91,14 @@ public ElementAttributes getElementInfos(UUID elementUuid) {
         return restTemplate.exchange(directoryServerBaseUri + path, HttpMethod.GET, null, ElementAttributes.class).getBody();
     }
 
-    private List<ElementAttributes> getElementsInfos(List<UUID> elementsUuids) {
+    private List<ElementAttributes> getElementsInfos(List<UUID> elementsUuids, List<String> elementTypes) {
         var ids = elementsUuids.stream().map(UUID::toString).collect(Collectors.joining(","));
         String path = UriComponentsBuilder.fromPath(ELEMENTS_SERVER_ROOT_PATH).toUriString() + "?ids=" + ids;
+
+        if (elementTypes != null) {
```
Comment (FranckLecuyer) [inline src/main/java/org/gridsuite/explore/server/services/DirectoryService.java:98]: ! CollectionUtils.isEmpty(elementTypes) ?
to avoid pb if elementTypes is not null but empty ...

File: src/main/java/org/gridsuite/explore/server/services/DirectoryService.java [hunk 2]
```diff
@@ -146,14 +151,22 @@ private IDirectoryElementsService getGenericService(String type) {
...
+        Map<String, List<ElementAttributes>> elementAttributesListByType = getElementsInfos(ids, elementTypes)
+                .stream()
                 .collect(Collectors.groupingBy(ElementAttributes::getType));
         List<ElementAttributes> listOfElements = new ArrayList<>();
         for (Map.Entry<String, List<ElementAttributes>> elementAttribute : elementAttributesListByType.entrySet()) {
             IDirectoryElementsService service = getGenericService(elementAttribute.getKey());
             listOfElements.addAll(service.completeElementAttribute(elementAttribute.getValue()));
         }
+
+        if (equipmentTypes != null && !equipmentTypes.isEmpty() && !listOfElements.isEmpty()) {
```
Comment (FranckLecuyer) [inline src/main/java/org/gridsuite/explore/server/services/DirectoryService.java:164]: ! CollectionUtils.isEmpty(equipmentTypes)  && !listOfElements.isEmpty() ?

File: src/main/java/org/gridsuite/explore/server/services/DirectoryService.java [hunk 3]
```diff
@@ -146,14 +151,22 @@ private IDirectoryElementsService getGenericService(String type) {
...
+            listOfElements = listOfElements.stream()
+                    .filter(element -> "DIRECTORY".equals(element.getType()) || elementTypes.contains(element.getSpecificMetadata().get("equipmentType")))
```
Comment (FranckLecuyer) [inline src/main/java/org/gridsuite/explore/server/services/DirectoryService.java:166]: equipmentTypes.contains(...) , instead of elementTypes.contains(...)

File: src/test/java/org/gridsuite/explore/server/ExploreTest.java [hunk 4]
```diff
@@ -413,6 +425,14 @@ public void testGetElementsMetadata() throws Exception {
...
                 .header("userId", USER1)
         ).andExpectAll(status().isOk());
+
+        mockMvc.perform(get("/v1/explore/elements/metadata?ids=" + FILTER_UUID + "," + FILTER_UUID_2 + "&elementTypes=FILTER")
+                .header("userId", USER1)
+        ).andExpectAll(status().isOk(), content().string("[{\"elementUuid\":\"" + FILTER_UUID + "\",\"elementName\":\"filterContingencyList\",\"type\":\"FILTER\",\"accessRights\":{\"private\":true},\"owner\":\"user1\",\"subdirectoriesCount\":0,\"description\":null,\"specificMetadata\":{\"id\":\"" + FILTER_UUID + "\"}},{\"elementUuid\":\"" + FILTER_UUID_2 + "\",\"elementName\":\"filterContingencyList2\",\"type\":\"FILTER\",\"accessRights\":{\"private\":true},\"owner\":\"user1\",\"subdirectoriesCount\":0,\"description\":null,\"specificMetadata\":{\"id\":\"" + FILTER_UUID_2 + "\"}}]"));
+
+        mockMvc.perform(get("/v1/explore/elements/metadata?ids=" + FILTER_UUID + "," + FILTER_UUID_2 + "&equipmentTypes=GENERATOR&elementTypes=FILTER")
+                .header("userId", USER1)
+        ).andExpectAll(status().isOk(), content().string("[]"));
```
Comment (FranckLecuyer) [inline src/test/java/org/gridsuite/explore/server/ExploreTest.java:435]: Add also a test with equipmentTypes=LINE, to verify that the response content is not empty in that case ...

File: src/main/java/org/gridsuite/explore/server/services/IDirectoryElementsService.java [hunk 5]
```diff
@@ -32,7 +32,8 @@ default List<ElementAttributes> completeElementAttribute(List<ElementAttributes>
         Map<String, ElementAttributes> mapElementAttribute = lstElementAttribute.stream()
                 .collect(Collectors.toMap(e -> e.getElementUuid().toString(), Function.identity()));
         /* getting metadata from services */
-        List<Map<String, Object>> metadata = getMetadata(lstElementAttribute.stream().map(ElementAttributes::getElementUuid).collect(Collectors.toList()));
+        List<UUID> collect = lstElementAttribute.stream().map(ElementAttributes::getElementUuid).collect(Collectors.toList());
+        List<Map<String, Object>> metadata = collect.stream().map(e -> Map.of("id", (Object) e)).collect(Collectors.toList());
```
Comment (FranckLecuyer) [inline src/main/java/org/gridsuite/explore/server/services/IDirectoryElementsService.java:36]: why the call of getMetadata method has disappeared ?

**Review by FranckLecuyer — CHANGES_REQUESTED** [review]

File: src/test/java/org/gridsuite/explore/server/ExploreTest.java (hunk 4 above)
Comment (YenguiSeddik) [reply src/test/java/org/gridsuite/explore/server/ExploreTest.java:443]: done

File: src/main/java/org/gridsuite/explore/server/services/DirectoryService.java (hunk 1 above)
Comment (YenguiSeddik) [reply src/main/java/org/gridsuite/explore/server/services/DirectoryService.java]: done

File: src/main/java/org/gridsuite/explore/server/services/DirectoryService.java (hunk 2 above)
Comment (YenguiSeddik) [reply src/main/java/org/gridsuite/explore/server/services/DirectoryService.java]: done

File: src/main/java/org/gridsuite/explore/server/services/DirectoryService.java (hunk 3 above)
Comment (YenguiSeddik) [reply src/main/java/org/gridsuite/explore/server/services/DirectoryService.java]: done

File: src/main/java/org/gridsuite/explore/server/services/IDirectoryElementsService.java (hunk 5 above)
Comment (YenguiSeddik) [reply src/main/java/org/gridsuite/explore/server/services/IDirectoryElementsService.java]: fixed

**Review by FranckLecuyer — APPROVED** [review]

File: src/test/java/org/gridsuite/explore/server/ExploreTest.java [hunk 6]
```diff
@@ -413,6 +433,18 @@ public void testGetElementsMetadata() throws Exception {
         mockMvc.perform(get("/v1/explore/elements/metadata?ids=" + FILTER_UUID + "," + PRIVATE_STUDY_UUID + "," + CONTINGENCY_LIST_UUID)
                 .header("userId", USER1)
         ).andExpectAll(status().isOk());
+
+        mockMvc.perform(get("/v1/explore/elements/metadata?ids=" + FILTER_UUID + "," + FILTER_UUID_2 + "&equipmentTypes=&elementTypes=FILTER")
+                .header("userId", USER1)
+        ).andExpectAll(status().isOk(), content().string("[{\"elementUuid\":\"" + FILTER_UUID + "\",\"elementName\":\"filterContingencyList\",\"type\":\"FILTER\",\"accessRights\":{\"private\":true},\"owner\":\"user1\",\"subdirectoriesCount\":0,\"description\":null,\"specificMetadata\":{\"id\":\"" + FILTER_UUID + "\"}},{\"elementUuid\":\"" + FILTER_UUID_2 + "\",\"elementName\":\"filterContingencyList2\",\"type\":\"FILTER\",\"accessRights\":{\"private\":true},\"owner\":\"user1\",\"subdirectoriesCount\":0,\"description\":null,\"specificMetadata\":{\"id\":\"" + FILTER_UUID_2 + "\",\"equipmentType\":\"LINE\"}}]"));
```
Comment (archendel) [inline src/test/java/org/gridsuite/explore/server/ExploreTest.java:439]: For all those expected response bodies, use "mapper.writeValueAsString(specificMetadata)" instead of writing the whole text ?

**Review by archendel — APPROVED** [review]
small comment for code readability

File: src/test/java/org/gridsuite/explore/server/ExploreTest.java (hunk 6 above)
Comment (YenguiSeddik) [reply src/test/java/org/gridsuite/explore/server/ExploreTest.java]: ok. done

**Review by archendel — APPROVED** [review]

**Review by FranckLecuyer — APPROVED** [review]
